File: y15/d08/cpp/main.cpp
#include "helper.h"
#include <iostream>
#include <ostream>
#include <vector>
#include <string>
using namespace std;

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  int total = 0;
  // Part 1
  int actual_total = 0;
  // Part 2
  int extra_total = 0;
  vector<string> lines = read_lines(argv[1]);
  for(const auto& line: lines) {
    int len = line.size();
    // Part 1
    int actual = 0;
    // Part 2
    int extra = 2;

    for (size_t i = 1; i < line.size() - 1;) {
      if(line[i] == '\\') {
        if (line[i+1] == '\\' || line[i + 1] == '"') {
          actual++;
          i += 2;
        } else if (line[i+1] == 'x'){
          actual++;
          i += 4;
        } else {
          actual++;
          i ++;
        }
      } else {
          actual++;
          i ++;
      }
    }

    total += len;

    actual_total += actual;

    for (size_t i = 0; i < line.size(); ) {
      if(line[i] == '\\' || line[i] == '"') {
        extra += 2;
      } else {
        extra ++;
      }

      i++;
    }

    extra_total += extra;
  }

  int part1 = total - actual_total;
  cout << "Part 1: " <<part1 << endl;

  int part2 = extra_total - total;
  cout << "Part 2: " <<part2 << endl;

  return 0;
}

File: y15/d08/rust/src/main.rs
use std::env;
use std::fs::read_to_string;

fn main() {
    let lines: Vec<String> = read_to_string(env::args().nth(1).expect("Error: Missing file name"))
        .unwrap()
        .lines()
        .map(String::from)
        .collect();
    solve(lines);
}

fn solve(lines: Vec<String>) {
    let mut total = 0;
    let mut actual_total = 0;
    let mut extra_total = 0;

    for line in lines {
        let len = line.len();
        let mut actual = 0;

        let chars: Vec<char> = line.chars().collect();
        let mut i = 1;
        while i < chars.len() - 1 {
            if chars[i] == '\\' {
                match chars[i + 1] {
                    '\\' | '"' => {
                        actual += 1;
                        i += 2;
                    }
                    'x' => {
                        actual += 1;
                        i += 4;
                    }
                    _ => {
                        actual += 1;
                        i += 1;
                    }
                }
            } else {
                actual += 1;
                i += 1;
            }
        }

        let mut i = 0;
        let mut extra = 2;
        while i < chars.len() {
            match chars[i] {
                '\\' | '"' => {
                    extra += 2;
                }
                _ => {
                    extra += 1;
                }
            }
            i += 1;
        }

        total += len;
        actual_total += actual;
        extra_total += extra
    }

    let part1 = total - actual_total;
    println!("Part 1: {}", part1);

    let part2 = extra_total - total;
    println!("Part 2: {}", part2)
}
